File: mcp-spike/java-spec/src/main/java/org/slizaa/mcp/javaspec/JavaKinds.java
package org.slizaa.mcp.javaspec;

import java.util.List;
import java.util.Set;

/**
 * Canonical constants for Java-specific node kinds, relationship kinds,
 * and group aliases used throughout Hierograph's tool surface.
 *
 * <p>All string constants match the namespaced vocabulary defined in the
 * tool surface proposal ({@code "java.class"}, {@code "java.method"}, etc.).
 * Use these constants instead of string literals to avoid typos and keep
 * the vocabulary in one place.</p>
 */
public final class JavaKinds {

    private JavaKinds() {}

    // ── node kinds ─────────────────────────────────────────────────────

    public static final String MODULE     = "java.module";
    public static final String PACKAGE    = "java.package";
    public static final String CLASS      = "java.class";
    public static final String INTERFACE  = "java.interface";
    public static final String ENUM       = "java.enum";
    public static final String RECORD     = "java.record";
    public static final String ANNOTATION = "java.annotation";
    public static final String METHOD     = "java.method";
    public static final String FIELD      = "java.field";
    public static final String PRIMITIVE  = "java.primitive";
    public static final String CONSTRUCTOR = "java.constructor";

    // ── kind groups ────────────────────────────────────────────────────

    /** All type-level node kinds. */
    public static final Set<String> TYPE_KINDS = Set.of(
            CLASS, INTERFACE, ENUM, RECORD, ANNOTATION
    );

    /** All member-level node kinds. */
    public static final Set<String> MEMBER_KINDS = Set.of(METHOD, FIELD);

    /** All valid node kinds (excluding pseudo-kinds like primitive and constructor). */
    public static final Set<String> ALL_KINDS = Set.of(
            MODULE, PACKAGE, CLASS, INTERFACE, ENUM, RECORD, ANNOTATION, METHOD, FIELD
    );

    // ── group aliases (accepted by kind_filter parameters) ─────────────

    public static final String ALIAS_TYPES    = "types";
    public static final String ALIAS_MEMBERS  = "members";
    public static final String ALIAS_PACKAGES = "packages";

    public static final Set<String> ALL_ALIASES = Set.of(
            ALIAS_TYPES, ALIAS_MEMBERS, ALIAS_PACKAGES
    );

    /**
     * Expands a group alias to its constituent kinds.
     * Returns {@code null} if the input is not a recognized alias.
     */
    public static List<String> expandAlias(String alias) {
        return switch (alias) {
            case ALIAS_TYPES    -> List.of(CLASS, INTERFACE, ENUM, RECORD, ANNOTATION);
            case ALIAS_MEMBERS  -> List.of(METHOD, FIELD);
            case ALIAS_PACKAGES -> List.of(PACKAGE);
            default -> null;
        };
    }

    // ── Java primitives ────────────────────────────────────────────────

    /** The set of Java primitive type names. */
    public static final Set<String> JAVA_PRIMITIVES = Set.of(
            "void", "boolean", "byte", "char", "short", "int", "long", "float", "double"
    );
}

File: mcp-spike/java-spec/src/main/kotlin/org/slizaa/mcp/javaspec/JavaKinds.kt
package org.slizaa.mcp.javaspec

/**
 * Canonical constants for Java-specific node kinds, relationship kinds,
 * and group aliases used throughout Hierograph's tool surface.
 *
 * All string constants match the namespaced vocabulary defined in the
 * tool surface proposal (`java.class`, `java.method`, etc.).
 * Use these constants instead of string literals to avoid typos and keep
 * the vocabulary in one place.
 */
object JavaKinds {

    // ── node kinds ─────────────────────────────────────────────────────

    const val MODULE = "java.module"
    const val PACKAGE = "java.package"
    const val CLASS = "java.class"
    const val INTERFACE = "java.interface"
    const val ENUM = "java.enum"
    const val RECORD = "java.record"
    const val ANNOTATION = "java.annotation"
    const val METHOD = "java.method"
    const val FIELD = "java.field"
    const val PRIMITIVE = "java.primitive"
    const val CONSTRUCTOR = "java.constructor"

    // ── kind groups ────────────────────────────────────────────────────

    /** All type-level node kinds. */
    @JvmField
    val TYPE_KINDS: Set<String> = setOf(CLASS, INTERFACE, ENUM, RECORD, ANNOTATION)

    /** All member-level node kinds. */
    @JvmField
    val MEMBER_KINDS: Set<String> = setOf(METHOD, FIELD)

    /** All valid node kinds (excluding pseudo-kinds like primitive and constructor). */
    @JvmField
    val ALL_KINDS: Set<String> = setOf(MODULE, PACKAGE, CLASS, INTERFACE, ENUM, RECORD, ANNOTATION, METHOD, FIELD)

    // ── group aliases (accepted by kind_filter parameters) ─────────────

    const val ALIAS_TYPES = "types"
    const val ALIAS_MEMBERS = "members"
    const val ALIAS_PACKAGES = "packages"

    @JvmField
    val ALL_ALIASES: Set<String> = setOf(ALIAS_TYPES, ALIAS_MEMBERS, ALIAS_PACKAGES)

    /**
     * Expands a group alias to its constituent kinds.
     * Returns `null` if the input is not a recognized alias.
     */
    @JvmStatic
    fun expandAlias(alias: String): List<String>? = when (alias) {
        ALIAS_TYPES -> listOf(CLASS, INTERFACE, ENUM, RECORD, ANNOTATION)
        ALIAS_MEMBERS -> listOf(METHOD, FIELD)
        ALIAS_PACKAGES -> listOf(PACKAGE)
        else -> null
    }

    // ── Java primitives ────────────────────────────────────────────────

    /** The set of Java primitive type names. */
    @JvmField
    val JAVA_PRIMITIVES: Set<String> = setOf(
        "void", "boolean", "byte", "char", "short", "int", "long", "float", "double"
    )
}
